File: 03-Stacks/20250726-beginner/ts/solution.test.ts
import { isBalanced } from './solution';

describe('isBalanced', () => {
    // Test Case 1: Empty string
    test('should return true for empty string', () => {
        expect(isBalanced('')).toBe(true);
    });

    // Test Case 2: Single pair of parentheses
    test('should return true for single pair of parentheses', () => {
        expect(isBalanced('()')).toBe(true);
    });

    // Test Case 3: Single pair of brackets
    test('should return true for single pair of brackets', () => {
        expect(isBalanced('[]')).toBe(true);
    });

    // Test Case 4: Single pair of braces
    test('should return true for single pair of braces', () => {
        expect(isBalanced('{}')).toBe(true);
    });

    // Test Case 5: Nested brackets of same type
    test('should return true for nested brackets of same type', () => {
        expect(isBalanced('((()))')).toBe(true);
    });

    // Test Case 6: Mixed balanced brackets
    test('should return true for mixed balanced brackets', () => {
        expect(isBalanced('([{}])')).toBe(true);
    });

    // Test Case 7: Unmatched opening bracket
    test('should return false for unmatched opening bracket', () => {
        expect(isBalanced('(()')).toBe(false);
    });

    // Test Case 8: Unmatched closing bracket
    test('should return false for unmatched closing bracket', () => {
        expect(isBalanced('())')).toBe(false);
    });

    // Test Case 9: Wrong order of brackets
    test('should return false for wrong order of brackets', () => {
        expect(isBalanced('([)]')).toBe(false);
    });

    // Test Case 10: Complex nested structure
    test('should return true for complex nested structure', () => {
        expect(isBalanced('{[(){}][(())]}')).toBe(true);
    });
});

File: 03-Stacks/20250726-beginner/go/solution_test.go
package main

import (
	"testing"
)

func TestIsBalanced(t *testing.T) {
	tests := []struct {
		name       string
		expression string
		expected   bool
	}{
		// Test Case 1: Empty string
		{
			name:       "Empty string",
			expression: "",
			expected:   true,
		},
		// Test Case 2: Single pair of parentheses
		{
			name:       "Single pair of parentheses",
			expression: "()",
			expected:   true,
		},
		// Test Case 3: Single pair of brackets
		{
			name:       "Single pair of brackets",
			expression: "[]",
			expected:   true,
		},
		// Test Case 4: Single pair of braces
		{
			name:       "Single pair of braces",
			expression: "{}",
			expected:   true,
		},
		// Test Case 5: Nested brackets of same type
		{
			name:       "Nested brackets of same type",
			expression: "((()))",
			expected:   true,
		},
		// Test Case 6: Mixed balanced brackets
		{
			name:       "Mixed balanced brackets",
			expression: "([{}])",
			expected:   true,
		},
		// Test Case 7: Unmatched opening bracket
		{
			name:       "Unmatched opening bracket",
			expression: "(()",
			expected:   false,
		},
		// Test Case 8: Unmatched closing bracket
		{
			name:       "Unmatched closing bracket",
			expression: "())",
			expected:   false,
		},
		// Test Case 9: Wrong order of brackets
		{
			name:       "Wrong order of brackets",
			expression: "([)]",
			expected:   false,
		},
		// Test Case 10: Complex nested structure
		{
			name:       "Complex nested structure",
			expression: "{[(){}][(())]}",
			expected:   true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := IsBalanced(tt.expression)
			if result != tt.expected {
				t.Errorf("IsBalanced(%q) = %v, expected %v", tt.expression, result, tt.expected)
			}
		})
	}
}